Add Plus Ones Broadcasting

rdobkins opened the pull request:
This adds files/makes changes to allow sharing/broadcasting of Plus Ones data to other group/raid members.  Also supports importing/exporting of Plus Ones data

papa-smurf commented on Classes/PlusOnes.lua:
```diff
@@ -2,126 +2,970 @@
+function PlusOnes:available()
+    GL:debug("PlusOnes:available");
+
+    --return GL:higherThanZero(DB:get("PlusOnes.MetaData.updatedAt", 0)); --not sure this check is even needed for plus ones
```
You want to keep this is place or want to add a `---@todo: ` and leave a comment for me so I can check it out later?

rdobkins replied:
Removed.  Allowed automatic importing without confirmation if not PlusOnes data was already available.  This way the confirmation dialog always shows regardless of whether the data is empty or not at the beginning of the import.  Also, prevented responding to a data request if no data was available.  This way it will broadcast an empty message which will prompt the receiver to clear their existing data.

papa-smurf commented on Classes/PlusOnes.lua:
```diff
@@ -2,126 +2,970 @@
+end
+
+--- Draw either the importer or overview
+--- based on the current boosted roll data
```
Looks like a find all for "boosted" is in order ^^

rdobkins replied:
Find all completed.  This was the only instance I found that I had missed.

papa-smurf commented on Classes/PlusOnes.lua:
```diff
@@ -2,126 +2,970 @@
+    local args = GL:strSplit(message, " ");
+
+    -- See if name is given.
+    if (#args > 1) then
```
No need to count `args` if all you care about is `args[2]` being present.
```lua
if (args[2]) then
```
Would suffice

rdobkins replied:
Fixed.

papa-smurf commented on Classes/PlusOnes.lua:
```diff
@@ -2,126 +2,970 @@
+        if (type(name) == "string"
+            and not GL:empty(name)
+        ) then
+            if (not DetailsByPlayerName[name]) then
```
Can just add `and not DetailsByPlayerName[name]` as a third check in the original if-statement

rdobkins replied:
Fixed.

papa-smurf commented on Classes/PlusOnes.lua:
```diff
@@ -2,126 +2,970 @@
+function PlusOnes:toPlusOne(plusOne)
+    plusOne = tonumber(plusOne);
+
+    if not plusOne then
```
Trivial: formatting (braces)

rdobkins replied:
I don't understand this comment.  Please explain.

phrdina replied:
I believe that you are supposed to use `if (not plusOne) then`.

papa-smurf replied:
Correct, could've/should've been more clear. Normally I change minor things like these instead of commenting on them, but that's not possible with PRs like these sadly.

Also keep in mind that quite some boosted roll logic was made by someone else, which definitely introduced inconsistencies

papa-smurf commented on Classes/PlusOnes.lua:
```diff
@@ -2,126 +2,970 @@
     self:triggerChangeEvent();
+end
+
+--[[--- Determine if a player is present in the table --- not sure if this is even called elsewhere
```
Temporary?

rdobkins replied:
Deleted the whole section.  I confirmed the function was not called.

papa-smurf commented on Classes/PlusOnes.lua:
```diff
@@ -2,126 +2,970 @@
+--- Get a player's plus one
+---
+---@param name string
+---@return void
```
Does not return void :)

rdobkins replied:
Fixed.

papa-smurf commented on Classes/PlusOnes.lua:
```diff
@@ -2,126 +2,970 @@
+    if (not self.MaterializedData.DetailsByPlayerName[normalizedName]) then
+        GL:tableSet(self.MaterializedData or {}, "DetailsByPlayerName." .. normalizedName .. ".total", plusOne);
+    else
+        self.MaterializedData.DetailsByPlayerName[normalizedName].total = plusOne;
```
If `self.MaterializedData` is not available in the `tableSet` above, then it will also fail here, because it still won't be set at this point

rdobkins replied:
self.MaterializedData is available because it is set in line 10.  Changed the code to remove the null table.

papa-smurf commented on Classes/PlusOnes.lua:
```diff
@@ -2,126 +2,970 @@
---- Get a player's PlusOnes value
+--- Delete an entry --- not sure where this would be needed since the player list is generated based on the group/raid members
+---
+---@param name string
```
param missing

rdobkins replied:
Fixed.  I need to add a ---@todo for the whole class to see if dontBroadcast is even needed.  I don't think the parameter is passed from any other function calls

papa-smurf commented on Classes/PlusOnes.lua:
```diff
@@ -2,126 +2,970 @@
 end
 
---- Get a player's PlusOnes value
+--- Delete an entry --- not sure where this would be needed since the player list is generated based on the group/raid members
```
Does this require me to look at it? If so add `---@todo` please

rdobkins replied:
Added a ---@todo to see if this function is even needed.

papa-smurf commented on Classes/PlusOnes.lua:
```diff
@@ -2,126 +2,970 @@
+    end
+end
+
+--- Modify a player's plus one -- don't know if this is needed since numbers are increased or decreased and not set directly
```
Same as above, should this stay in then?

rdobkins replied:
Same comment as above.

papa-smurf commented on Classes/PlusOnes.lua:
```diff
@@ -2,126 +2,970 @@
+--- Export to CSV
+---
+---@param displayFrame boolean
+---@return void
```
Return type incorrect

rdobkins replied:
I used the same return type you used in Classes\BoostedRolls.lua

papa-smurf commented on Classes/PlusOnes.lua:
```diff
@@ -2,126 +2,970 @@
+        if (GL:empty(importString)) then
+            self:clearPlusOnes();
+            return;
+            --GL:warning("Couldn't process PlusOnes data received from " .. CommMessage.Sender.name);
```
What happens if the broadcast is empty?

rdobkins replied:
What happens in the case of Plus Ones, is it allows clearing of all data when the ML or assist clears the plus one data.  For example, if I do multiple raids in one night and I want to clear the plus ones between the raids (i.e. do EoE then head to OS).  The commented section has been removed.

papa-smurf commented on Classes/PlusOnes.lua:
```diff
@@ -2,126 +2,970 @@
+    end
 
-        DB.PlusOnes[playerName] = GL:round(value or 0);
+    --if (not self:userIsAllowedToBroadcast()) then
```
You sure we don't want to put this in place?

rdobkins replied:
PlusOnes database is now similar to Boosted Rolls.  The PlusOnes data is accessed as DB.PlusOnes.Totals[playerName].total  Commented section deleted.  This is implemented in the following code without the warning message.  That way if anyone wants to maintain their own list they won't get bombarded with warning messages if they inadvertently have sharing enabled but don't have permissions.

papa-smurf reviewed (changes requested):
Pretty good for a first PR! I see some discrepancies that were copy/pasted from boosted roll logic or other places that are outdated to begin with, so that's definitely understandable.

This sadly is way too much at once for me to just merge and it will require a lot of testing and double-checking on my part on both classic and retail, which is also the reason why it will take a bit before I can free up enough time to fully process it I hope that's understandable.

I do really appreciate the effort you put in, let's stay in touch!

rdobkins commented:
I don't see a conflict with Data/Constants.lua anymore.

papa-smurf commented:
@rdobkins while testing I like to just keep track of everything I encounter in paint so I can include visuals easily, hope that works for you:

![image](https://user-images.githubusercontent.com/4224975/211158949-cb1fbe31-ea18-4f35-a18f-468afac04f76.png)

rdobkins commented:
> @rdobkins while testing I like to just keep track of everything I encounter in paint so I can include visuals easily, hope that works for you:

Paint works fine for me.

Import button location fixed.  Was an ElvUI formatting error.

I felt overall that only seeing the +1s for people in the group would be preferable.  The class could be rewritten to be like boosted rolls where a static list is maintained and raiders have to be manually added and deleted.

Not sure which two buttons you are showing but to answer the comment: The intent is that the raid leader's value changes regardless of whether the receiver accepts or not.  The receiver's value should not update unless they accept the broadcast.  That way their list can be unchanged.

There should be a popup for every broadcast unless the leader has been added to the allowed list.  There shouldn't be any dialogs without context unless I missed something:  Attached are broadcast and change dialogs with receiver results.  Initial ML values are 3 and 2 before initial broadcast and 3 and 3 after 1st change:

![Master Looter Broadcast Accept Dialog](https://user-images.githubusercontent.com/52630237/211163501-0e761d2b-1c8b-4b57-b45c-4e3609fb4a54.jpg)
![Master Looter Change Dialog Pre Accept](https://user-images.githubusercontent.com/52630237/211163508-42c76f3e-2531-44f1-b3c7-896209ea0881.jpg)
![Master Looter Change Post Accept](https://user-images.githubusercontent.com/52630237/211163519-172acfc4-1ba2-4c9b-8fed-3c9d02ece441.jpg)

I agree +1 vice PO makes more sense and that's an easy fix.

Right now the "Enable Plus Ones" setting only affects the award dialog where the checkboxes for +1 and OS show up.  If 'Plus Ones" are NOT enabled only the OS checkbox shows up

![Award Dialog with Plus Ones Enabled](https://user-images.githubusercontent.com/52630237/211160526-94dfb07d-4f3f-4401-a0c9-800bbcf589fb.jpg)

![Award Dialog with Plus Ones Disabled](https://user-images.githubusercontent.com/52630237/211160559-0cf49685-c2ce-4706-ba91-74244b3cd7a1.jpg)

Added listener for GL.GROUP_ROSTER_CHANGED_THROTTLED to close and redraw window when group changes.

rdobkins commented:
One of the things that is confusing is that the imported at time gets set to current server time if no import has ever occurred.  So it looks weird when you open the window some time after the last update and the imported at time is newer when no import has occurred.
